File: src/client/perl/p4/jira-client.pl
#!/usr/bin/perl

# JIRA commit acceptance perl client for P4
# Author: ferenc.kiss@midori.hu
# $Id: jira-client.pl 10144 2007-09-12 10:13:37Z ferenc.kiss $

use strict;
use warnings;

use XMLRPC::Lite;

# configure JIRA access
# ($projectKey can contain multiple comma-separated JIRA project keys like 'my $projectKey = "TST,ARP";'.
# If you specify multiple keys, the commit will be accepted if at least one project listed accepts it.
# Or you can specify 'my $projectKey = "*";' to force using the global commit acceptance settings if you don't
# want to specify any exact project key.)
my $jiraBaseURL = "http://127.0.0.1:8083";
my $jiraLogin = "admin";
my $jiraPassword = "admin";
my $projectKey = "TST";

# configure p4 path
my $p4Path = "p4";

# get committer passed as arg[0]
my $committer = $ARGV[0];

# get change description with change passed as arg[1]
my $holdTerminator = $/;
undef $/;

open IN, '-|', "$p4Path -s describe $ARGV[1]" or die "Unable to get change description with p4.\n";
my $buffer = <IN>;
close IN;

$/ = $holdTerminator;
my @lines = split /$holdTerminator/, $buffer;
@lines = map(substr($_, 1), grep(/^\t/, map(substr($_, 6) , grep(/^text:/, @lines)))); # TODO review and comment
$buffer = "init";
$buffer = join $holdTerminator, @lines;
chomp($buffer);

my $commitMessage = $buffer;

# print arguments
print "Committer: " . $committer . "\n";
print "Commit message: \"" . $commitMessage . "\"\n";

# invoke JIRA web service
# TODO move this common  part to a separate.pl and use it also from the SVN client,
my $acceptance;
my $comment;

eval {
	$jiraBaseURL =~ s/\/+$//; # remove trailing '/' if exists
	my $s = XMLRPC::Lite->proxy($jiraBaseURL . "/rpc/xmlrpc");
	my $result = $s->call("commitacc.acceptCommit", $jiraLogin, $jiraPassword, $committer, $projectKey, $commitMessage)->result();
	($acceptance, $comment) = split('\|', $result);
};

if ($@) {
	($acceptance, $comment) = ("false", "Unable to connect to the JIRA server at \"" . $jiraBaseURL . "\".");
};

if($acceptance eq 'true') {
	print "Commit accepted.\n";
	exit 0;
} else {
	print "Commit rejected: " . $comment . "\n";
	exit 1;
}

File: src/client/python/p4/jira-client.py
#!/usr/bin/python

# JIRA commit acceptance python client for P4
# Author: ferenc.kiss@midori.hu
# $Id: jira-client.py 10144 2007-09-12 10:13:37Z ferenc.kiss $

import os
import sys
import urlparse
import xmlrpclib

# configure JIRA access
# ("projectKey" can contain multiple comma-separated JIRA project keys like "projectKey = 'TST,ARP'".
# If you specify multiple keys, the commit will be accepted if at least one project listed accepts it.
# Or you can specify "projectKey = '*'" to force using the global commit acceptance settings if you don't
# want to specify any exact project key.)
jiraBaseURL = 'http://127.0.0.1:8080'
jiraLogin = 'admin'
jiraPassword = 'admin'
projectKey = 'TST'

# configure svnlook path
p4Path = 'p4'

# get committer passed as arg[1]
committer = sys.argv[1]

# get change description with change passed as arg[2]
try:
	f = os.popen(p4Path + ' -s describe ' + sys.argv[2])
	commitMessage = f.read()
	if f.close():
		raise 1
except:
	print >> sys.stderr, 'Unable to get change <' + sys.argv[2] + '> description with p4.'
	sys.exit(1)

# TODO trim leading whitespace
	
# print arguments
print 'Committer: ' + committer
print 'Commit message: "' + commitMessage + '"'

# invoke JIRA web service
try:
	s = xmlrpclib.ServerProxy(urlparse.urljoin(jiraBaseURL, '/rpc/xmlrpc'))
	acceptance, comment = s.commitacc.acceptCommit(jiraLogin, jiraPassword, committer, projectKey, commitMessage).split('|');
except:
	acceptance, comment = ['false', 'Unable to connect to the JIRA server at "' + jiraBaseURL + '".']

if acceptance == 'true':
	print 'Commit accepted.'
	sys.exit(0)
else:
	print 'Commit rejected: ' + comment
	sys.exit(1)
